### ISSUE TITLE
Bug: [no-base-to-string] check array generic type

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#10430]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=MoFwTglgdg5gFAbQN4F8A0ACVBdAlAbgwHoiMByBAewCMArAUwGMQMB5OpkbNKj5tvlzJA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oCMBDZRLXxdk%2BKkwDm6KAHde0JpHBgAviBVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
String([{}, {}]); // '[object Object],[object Object]'
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-base-to-string": "warn"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect a lint report

### Actual Result

no report

### Additional Info

The provided code is effectively identical to 
```ts
[{}, {}].join(','); // '[object Object],[object Object]'
```
which does flag, since https://github.com/typescript-eslint/typescript-eslint/issues/3388

**Comment by kirkwaiblinger** [comment]
Code-wise, this is just a matter of reusing the same `isTuple()` and `isArray()` checks in the array join case.
